File: 2903-insert-greatest-common-divisors-in-linked-list/2903-insert-greatest-common-divisors-in-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* insertGreatestCommonDivisors(ListNode* head) {
        auto gcd = [](int a, int b) {
            while (b != 0) {
                int temp = b;
                b = a % b; // Remainder becomes the new b
                a = temp;  // Old b becomes the new a
            }
            return a;
        };

        ListNode* temp = head;
        while (temp->next) {
            ListNode* node = new ListNode(gcd(temp->val, temp->next->val), temp->next);
            temp->next = node;
            temp = node->next;
        }
        return head;
    }
};

File: 2903-insert-greatest-common-divisors-in-linked-list/2903-insert-greatest-common-divisors-in-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
      
    def insertGreatestCommonDivisors(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # def gcd(a, b):
        #     while b:
        #         a, b = b, a%b
        #     return a

        temp = head
        while temp.next:
            l = temp
            r = temp.next
            m = ListNode(math.gcd(l.val, r.val))
            l.next = m
            m.next = r
            temp = r
        return head




        